File: rest/src/main/java/io/disquark/rest/resources/partial/PartialAttachment.java
package io.disquark.rest.resources.partial;

import java.util.Optional;
import java.util.OptionalInt;

import com.fasterxml.jackson.annotation.JsonProperty;

import io.disquark.immutables.ImmutableJson;
import io.disquark.rest.resources.Snowflake;

@ImmutableJson
public interface PartialAttachment {

    static Builder builder() {
        return new Builder();
    }

    Snowflake id();

    Optional<String> filename();

    Optional<String> description();

    @JsonProperty("content_type")
    Optional<String> contentType();

    OptionalInt size();

    Optional<String> url();

    @JsonProperty("proxy_url")
    Optional<String> proxyUrl();

    OptionalInt height();

    OptionalInt width();

    Optional<Boolean> ephemeral();

    class Builder extends ImmutablePartialAttachment.Builder {
        protected Builder() {
        }
    }
}

File: rest/kotlin/src/main/kotlin/json/message/PartialAttachment.kt
package io.disquark.rest.kotlin.json.message

import io.disquark.rest.json.Snowflake
import io.disquark.rest.json.message.PartialAttachment as ImmutablePartialAttachment

data class PartialAttachment(val id: Snowflake, var filename: String? = null, var description: String? = null) {
    internal fun toImmutable(): ImmutablePartialAttachment {
        return ImmutablePartialAttachment(id)
            .run { filename?.let { withFilename(it) } ?: this }
            .run { description?.let { withDescription(it) } ?: this }
    }
}
